url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/unpacker/Converter.java

```diff
@@ -0,0 +1,438 @@
+                    "Expected array but got not array value");
+        }
+        ArrayValue a = v.asArrayValue();
+        stack.reduceCount();
+        stack.pushArray(a.size());
+        values[stack.getDepth()] = a.getElementArray();
+        return a.size();
+    }
+
+    
+    public void readArrayEnd(boolean check) throws IOException {
+        if (!stack.topIsArray()) {
+            throw new MessageTypeException(
+                    "readArrayEnd() is called but readArrayBegin() is not called");
+        }
+
+        int remain = stack.getTopCount();
+        if (remain > 0) {
+            if (check) {
+                throw new MessageTypeException(
+                        "readArrayEnd(check=true) is called but the array is not end");
+            }
+            for (int i = 0; i < remain; i++) {
+                skip();
+            }
+        }
+        stack.pop();
+
+        if (stack.getDepth() == 0) {
+            value = null;
+        }
+    }
+
+    
+    public int readMapBegin() throws IOException {
+        Value v = getTop();
+        if (!v.isMapValue()) {
+            throw new MessageTypeException("Expected map but got not map value");
+        }
+        MapValue m = v.asMapValue();
+        stack.reduceCount();
+        stack.pushMap(m.size());
+        values[stack.getDepth()] = m.getKeyValueArray();
+        return m.size();
+    }
+
+    
+    public void readMapEnd(boolean check) throws IOException {
+        if (!stack.topIsMap()) {
+            throw new MessageTypeException(
+                    "readMapEnd() is called but readMapBegin() is not called");
+        }
+
+        int remain = stack.getTopCount();
+        if (remain > 0) {
+            if (check) {
+                throw new MessageTypeException(
+                        "readMapEnd(check=true) is called but the map is not end");
+            }
+            for (int i = 0; i < remain; i++) {
+                skip();
+            }
+        }
+        stack.pop();
+
+        if (stack.getDepth() == 0) {
+            value = null;
+        }
+    }
+
+    private Value getTop() throws IOException {
+        ensureValue();
+
+        stack.checkCount();
+        if (stack.getDepth() == 0) {
+            // if(stack.getTopCount() < 0) {
+            // //throw new EOFException(); // TODO
+            // throw new RuntimeException(new EOFException());
+            // }
+            return value;
+        }
+        Value[] array = (Value[]) values[stack.getDepth()];
+        return array[array.length - stack.getTopCount()];
+    }
+
+    
+    public Value readValue() throws IOException {
+        if (stack.getDepth() == 0) {
+            if (value == null) {
+                return nextValue();
+            } else {
+                Value v = value;
+                value = null;
+                return v;
+            }
+        }
+        return super.readValue();
+    }
+
+    
+    protected void readValue(Unconverter uc) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 309-319 into a separate method.